### ISSUE TITLE
Use Multipart upload API to upload files larger than 5 GB

### ISSUE THREAD

**Issue by ptirador** [carlspring/s3fs-nio#95]
# Task Description
The multipart upload API is designed to improve the upload experience for larger objects. You can upload an object in parts. These object parts can be uploaded independently, in any order, and in parallel. You can use a multipart upload for objects from 5 MB to 5 TB in size.

# Tasks

The following tasks will need to be carried out:
* [ ] First create a multipart upload with `CreateMultipartUploadRequest` and get the upload id.
* [ ] Upload all the different parts of the object with the help of `UploadPartRequest` and `CompletedPart`.
* [ ] Finally call `completeMultipartUpload` operation with `CompletedMultipartUpload` request to tell S3 to merge all uploaded parts and finish the multipart operation.

# Task Relationships

This task:
* Is a follow-up of: #63

# Useful Links
* [Uploading objects](https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html)
* [Amazon S3 multipart upload limits](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html)
* [Upload objects in multiple parts](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3-objects.html#list-objects)
* [Pull Request on Upplication/Amazon-S3-FileSystem-NIO2 (SDK v1)](https://github.com/Upplication/Amazon-S3-FileSystem-NIO2/pull/45)

# Help

* Our [chat channel](https://chat.carlspring.org/channel/s3fs-nio-community)
* Points of contact:
  * @carlspring
  * @steve-todorov
  * @sbespalov

**Comment by ptirador** [comment]
There are 2 kinds of APIs to perform this multi-part operation: high-level API and low-level API:
We should use the low-level API when we need to pause and resume multipart uploads, vary part sizes during the upload, or do not know the size of the upload data in advance. When we don't have these requirements, we should use the high-level API.

If we upload files in the S3 bucket using `TransferManager`(high-level API), which it's easier to integrate, we can upload an  `InputStream` or `File`, and also we can use multipart upload in `TransferManager` by configuring `MultipartUploadThreshold` in `TransferManager` client. But for this client we need to provide the content-length of the file size in the `PutObjectRequest` before uploading the file to the S3 bucket. And when dealing with `InputStream` of a large file we might not get the content-length in the HTTP Response, so in that case, we cannot use `TransferManager` (high-level API).

So, in summary, I think low-level API seems more flexible and manageable, although it may be not so easy to integrate, but in the long run it is worth it.

cc @carlspring

**Comment by carlspring** [comment]
Hi,

Thanks for having a preliminary look, @ptirador! I don't know, if you've seen the existing pull requests in the Upplication upstream and whether there's anything we could use as inspiration from there?

@elerc , @markjschreiber,

Do you guys have any suggestions and recommendations?
